baybe/searchspace/discrete.py

```diff
@@ -710,6 +739,79 @@ def parameter_cartesian_prod_to_df(
     return ret
 
 
+def parameter_cartesian_prod_polars_with_filtering(
+    parameters: Iterable[Parameter], constraints: Iterable[DiscreteConstraint]
+) -> pd.DataFrame:
+    """Create the Cartesian product of all parameter values.
+
+    Ignores continuous parameters.
+
+    Args:
+        parameters: List of parameter objects.
+        constraints: List of constraint objects.
+
+    Returns:
+        A dataframe containing all possible discrete parameter value combinations.
+    """
+    discrete_parameters = [p for p in parameters if isinstance(p, DiscreteParameter)]
+    if not discrete_parameters:
+        return pd.DataFrame()
+
+    param_frames = [pl.LazyFrame({p.name: p.values}) for p in discrete_parameters]
+    if len(param_frames) < 1:
+        return pd.DataFrame()
+
+    if len(param_frames) == 1:
+        return param_frames[0].collect().to_pandas()
+
+    constraints = [c for c in constraints if hasattr(c, "get_polars_expression")]
+
+    res = param_frames[0]
+
+    for frame in param_frames[1:]:
+        # Join first
+        res = res.join(frame, how="cross", force_parallel=True)
+        # Filter constraints
+        for c in constraints:
+            pl_expr = c.get_polars_expression(res)  # type: ignore[attr-defined]
+            if pl_expr is not None:
+                res = res.filter(pl_expr)
+
+    return res.collect(streaming=True).to_pandas()
+
+
+def parameter_cartesian_prod_polars(parameters: Iterable[Parameter]) -> pl.LazyFrame:
+    """Create the Cartesian product of all parameter values.
+
+    Ignores continuous parameters.
+
+    Args:
+        parameters: List of parameter objects.
+
+    Returns:
+        A lazy dataframe containing all possible discrete parameter value combinations.
+    """
+    discrete_parameters = [p for p in parameters if isinstance(p, DiscreteParameter)]
+    if not discrete_parameters:
+        return pl.LazyFrame()
+
+    # Convert each parameter to a lazy dataframe for cross-join operation
+    param_frames = [pl.LazyFrame({p.name: p.values}) for p in discrete_parameters]
+
+    # Handling edge cases
+    if len(param_frames) < 1:
```

Comment:
this if block shouldnt be possible anymore after your already made sure that `discrete_parameters` is not empty above?

Reply:
Yeah that's right😅